scripts/metadata_enrichment/tob_bioheart_sequencing_dates.py

```diff
@@ -0,0 +1,247 @@
+    ).to_dict()
+
+
+# TODO: combine api calls and
+async def upsert_sequencing_dates(
+    fluidx_to_assay_ids: defaultdict,
+    fluidx_to_sequencing_date: dict,
+):
+    # construct API update calls
+    # Iterate through the fluidX_to_assay_IDs dictionary because this is representative of what's already in metamist
+    # That is: fluidX_to_assay_IDs groups assays by fluidX_tube_id
+
+    assay_api = AssayApi()
+    assays_to_update = []
+
+    for fluidx_id, assay_ids in fluidx_to_assay_ids.items():
+        # TODO: Refactor with early return as suggested in PR
+        # Get sequencing date for each fluidX id from fluidX_to_sequencing_date dict
+        if not fluidx_id:
+            print(f'*****\nNO TUBE ID FOR : {fluidx_id} and assays {assay_ids}')
+            continue
+
+        try:
+            # Create object required to update assay
+            date = fluidx_to_sequencing_date[fluidx_id]
+        except KeyError:
+            print(f'Tube ID {fluidx_id} is not in the sequencing manifest')
+        else:
+            for assay_id in assay_ids:
+                updated_assay = AssayUpsert(
+                    id=assay_id,
+                    meta={'fluid_x_tube_sequencing_date': date},
+                )
+                assays_to_update.append(
+                    assay_api.update_assay_async(assay_upsert=updated_assay),
+                )
+
+    return await asyncio.gather(*assays_to_update)
+
+
+def compare_tubes_metamist_excel(
+    fluidx_to_assay_ids: defaultdict,
+    fluidx_to_sequencing_date: dict,
+):
+    """
+    Includes diagnostic code used to identify tubes missing between metamist and the
+    provided sequencing date manifests. Prints output to console.
+    """
+    # Create a set from the fluidX identifiers only that were extracted from metamist
+    metamist_set_fluidx = set(fluidx_to_assay_ids.keys())
+
+    # Create a second set with fluidX identifiers extracted from excel files
+    excel_set_fluidx = set(fluidx_to_sequencing_date.keys())
+
+    # Find intersection/difference between two sets
+    print(f'Count fluidX in Metamist {len(metamist_set_fluidx)}')
+    print(f'Count fluidX in Excel {len(excel_set_fluidx)}')
+
+    diff_metamist_excel = metamist_set_fluidx.difference(excel_set_fluidx)
+    print(f'Diff metamist excel {diff_metamist_excel}')
+    print(f'Len diff metamist excel {len(diff_metamist_excel)}')
+    diff_excel_metamist = excel_set_fluidx.difference(metamist_set_fluidx)
+    print(f'Diff excel metamist {diff_excel_metamist}')
+    print(f'Len diff excel metamist {len(diff_excel_metamist)}')
+
+    # Which assays are associated with the 'None' tube?
+    assays_no_tubes = fluidx_to_assay_ids.get(None, 'empty')
+    print(assays_no_tubes)
+    print(len(assays_no_tubes[0]))
+
+
+async def main():
+    fluidx_to_assay_ids = query_metamist()
+    if len(fluidx_to_assay_ids) != 0:
+        fluidx_to_sequencing_date = extract_excel()
+        await upsert_sequencing_dates(
```

Comment:
I think you could add a commented out line or two here like:
```python
# Uncomment this to compare the data in metamist to the manifests
# compare_tubes_metamist_excel(fluidx_to_assay_ids, fluidx_to_sequencing_date)
```

Because currently the compare_tubes_metamist_excel function is unused, which is fine, the docstring explains it's there for debugging, but it might be good to show where in the code it would be used.

You could go even further and add a click option like `@click.option('--debug', '-d', is_flag=True)`
And then in the `main` function:
```python
if debug:
   compare_tubes_metamist_excel(fluidx_to_assay_ids, fluidx_to_sequencing_date)
```

Refactoring and adding click isn't necessary, but still I think it would be good to include the function somewhere, even if it's commented out for the purposes of running the script on production data.

Reply:
Have added the two comments as suggested.

I agree that `click` makes this more reusable in the longer term but I'm going to leave this as-is for now.

Reply:
I changed my mind and added the click module 👯